### ISSUE TITLE
Samtools sort default temporary location is error-prone

### ISSUE THREAD

**Issue by smowton** [samtools/samtools#535]
Hi,

I ran two invocations of samtools sort in parallel, both of which were reading from standard in (specified by passing '-' as the input argument). I was surprised to find they failed because they both tried to store temporary files at -.tmp.nnnn.bam and overwrote each other's results. Similarly it looks from bam_sort.c:1810 as if STDIN.tmp.nnnn.bam might be a similar subject of contention if I hadn't passed a positional argument.

Since writing related to the input filename is likely to fail anyhow (e.g. the input filesystem or directory might be read-only), I suggest you should default to placing temporaries next to the *output* filename if you have one, since concurrently writing to the same output is always an error, or generating a fresh name in $TMPDIR if you're writing to stdout.

**Comment by jmarshall** [comment]
Duplicate of #523.
